ebos/eclcpgridvanguard.hh

```diff
@@ -189,17 +191,29 @@ public:
             //distribute the grid and switch to the distributed view.
             {
                 const auto wells = this->schedule().getWellsatEnd();
-                auto& eclState = static_cast<ParallelEclipseState&>(this->eclState());
-                const EclipseGrid* eclGrid = nullptr;
 
-                if (grid_->comm().rank() == 0)
+                try
                 {
-                    eclGrid = &this->eclState().getInputGrid();
-                }
+                    auto& eclState = dynamic_cast<ParallelEclipseState&>(this->eclState());
+                    const EclipseGrid* eclGrid = nullptr;
+
+                    if (grid_->comm().rank() == 0)
+                    {
+                        eclGrid = &this->eclState().getInputGrid();
+                    }
 
-                PropsCentroidsDataHandle<Dune::CpGrid> handle(*grid_, eclState, eclGrid, this->centroids_,
-                                                              cartesianIndexMapper());
-                defunctWellNames_ = std::get<1>(grid_->loadBalance(handle, edgeWeightsMethod, &wells, faceTrans.data()));
+                    PropsCentroidsDataHandle<Dune::CpGrid> handle(*grid_, eclState, eclGrid, this->centroids_,
+                                                                  cartesianIndexMapper());
+                    defunctWellNames_ = std::get<1>(grid_->loadBalance(handle, edgeWeightsMethod, &wells, faceTrans.data()));
+                }
+                catch(std::bad_cast e)
```

Comment:
catch by ref

Reply:
mmh. I know "catch by reference and throw by value". To satisfy both I would need to make a copy anyway. Shouldn't the copy in the catch clause be elided anyway (probably still not guaranteed)?

Summoning @bska.

Reply:
thing is std::bad_cast is polymorphic, so you get a warning, https://ci.opm-project.org/job/opm-simulators-PR-builder/746/gcc/new/source.9c010f09-9839-4bb1-807a-2db5201de9aa/#245

Reply:
ok, I am convinced.